File: context_os/llm/anthropic_client.py
"""Anthropic Claude LLM 客户端实现。"""

from __future__ import annotations

import json
import os
from typing import Any, Optional

from anthropic import AsyncAnthropic

from context_os.core.logger import get_logger
from context_os.llm.client import BaseLLMClient

logger = get_logger(__name__)


class AnthropicClient(BaseLLMClient):
    """Anthropic Claude API 客户端。

    Args:
        api_key: API 密钥，默认从 ANTHROPIC_API_KEY 环境变量读取。
        model: 模型名称，默认 claude-sonnet-4-20250514。
    """

    def __init__(
        self,
        api_key: Optional[str] = None,
        model: str = "claude-sonnet-4-20250514",
    ):
        self.api_key = api_key or os.environ.get("ANTHROPIC_API_KEY", "")
        if not self.api_key:
            logger.warning("ANTHROPIC_API_KEY not set")
        self.model = model
        self._client = AsyncAnthropic(api_key=self.api_key)
        logger.info("AnthropicClient initialized: model=%s", model)

    async def complete(
        self,
        prompt: str,
        system: Optional[str] = None,
        max_tokens: int = 4096,
        temperature: float = 0.7,
        response_format: Optional[str] = None,
    ) -> str | dict[str, Any]:
        """发送请求到 Claude。

        Args:
            prompt: 用户 prompt。
            system: 系统指令。
            max_tokens: 最大输出 token 数。
            temperature: 温度参数。
            response_format: "json" 或 None。

        Returns:
            字符串或解析后的 JSON 字典。
        """
        logger.debug(
            "Claude request: model=%s, max_tokens=%d, system=%s",
            self.model, max_tokens, bool(system),
        )

        kwargs = {
            "model": self.model,
            "max_tokens": max_tokens,
            "temperature": temperature,
            "messages": [{"role": "user", "content": prompt}],
        }
        if system:
            kwargs["system"] = system

        try:
            response = await self._client.messages.create(**kwargs)
            text = response.content[0].text

            logger.debug(
                "Claude response: stop_reason=%s, input_tokens=%d, output_tokens=%d",
                response.stop_reason,
                response.usage.input_tokens,
                response.usage.output_tokens,
            )

            if response_format == "json":
                return json.loads(text)
            return text

        except Exception as e:
            logger.error("Claude API error: %s", e, exc_info=True)
            raise

File: java-context-os/src/main/java/com/owencli/contextos/llm/AnthropicClient.java
package com.owencli.contextos.llm;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.owencli.contextos.core.base.BaseLLMClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;

/**
 * Anthropic Claude API client.
 */
public class AnthropicClient implements BaseLLMClient {

    private static final Logger log = LoggerFactory.getLogger(AnthropicClient.class);
    private static final String DEFAULT_BASE_URL = "https://api.anthropic.com/v1";
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final String apiKey;
    private final String model;
    private final String baseUrl;
    private final HttpClient httpClient;

    public AnthropicClient() {
        this(
                System.getenv("ANTHROPIC_API_KEY"),
                "claude-sonnet-4-20250514",
                DEFAULT_BASE_URL
        );
    }

    public AnthropicClient(String apiKey, String model, String baseUrl) {
        this.apiKey = apiKey != null && !apiKey.isEmpty() ? apiKey : System.getenv("ANTHROPIC_API_KEY");
        this.model = model != null ? model : "claude-sonnet-4-20250514";
        this.baseUrl = baseUrl != null ? baseUrl : DEFAULT_BASE_URL;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .build();

        if (this.apiKey == null || this.apiKey.isEmpty()) {
            log.warn("ANTHROPIC_API_KEY not set");
        }
        log.info("AnthropicClient initialized: model={}", this.model);
    }

    @Override
    public CompletableFuture<Object> complete(String prompt, String system, int maxTokens,
                                              double temperature, String responseFormat) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                var root = MAPPER.createObjectNode();
                root.put("model", model);
                root.put("max_tokens", maxTokens);
                root.put("temperature", temperature);

                if (system != null && !system.isEmpty()) {
                    root.put("system", system);
                }

                ArrayNode messages = MAPPER.createArrayNode();
                var msg = MAPPER.createObjectNode();
                msg.put("role", "user");
                msg.put("content", prompt);
                messages.add(msg);
                root.set("messages", messages);

                String requestBody = MAPPER.writeValueAsString(root);
                log.debug("Claude request: model={}, max_tokens={}", model, maxTokens);

                var request = HttpRequest.newBuilder()
                        .uri(URI.create(baseUrl + "/messages"))
                        .header("x-api-key", apiKey)
                        .header("anthropic-version", "2023-06-01")
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                        .timeout(Duration.ofSeconds(120))
                        .build();

                HttpResponse<String> response = httpClient.send(request,
                        HttpResponse.BodyHandlers.ofString());

                if (response.statusCode() != 200) {
                    log.error("Claude API error: status={}, body={}",
                            response.statusCode(), response.body());
                    throw new RuntimeException("Claude API error: " + response.statusCode());
                }

                JsonNode json = MAPPER.readTree(response.body());
                String text = json.get("content").get(0).get("text").asText();

                log.debug("Claude response: stop_reason={}",
                        json.get("stop_reason").asText());

                if ("json".equals(responseFormat)) {
                    return MAPPER.readTree(text);
                }
                return text;

            } catch (Exception e) {
                log.error("Claude API error: {}", e.getMessage(), e);
                throw new RuntimeException("Claude API call failed", e);
            }
        });
    }
}
